io-amqp1_0-impl/src/main/java/org/apache/pulsar/ecosystem/io/amqp/AmqpSource.java

```diff
@@ -111,17 +114,38 @@ public void onMessage(Message message) {
     private static class QpidJmsRecord implements Record<ByteBuffer> {
         private final Optional<String> key;
         private final ByteBuffer value;
+        private final JmsMessage jmsMessage;
 
         @Override
         public Schema<ByteBuffer> getSchema() {
             return Schema.BYTEBUFFER;
         }
+
+        @Override
+        public void ack() {
+            try {
+                jmsMessage.acknowledge();
+            } catch (JMSException e) {
+                log.error("Failed to acknowledge qpid jms message.", e);
+            }
+        }
+
+        @Override
+        public void fail() {
+            try {
+                if (jmsMessage.getAcknowledgeCallback() != null) {
```

Comment:
If the acknowledge callback is null, will the message be still rejected?

Reply:
It is not acked, so in the end (depending on the settings of the source, in our case an AMQ Broker) a timeout will result in a nack of the message.

Reply:
Ok, thanks for your explanation.